pkg/tinydtls/contrib/sock_dtls.c

```diff
@@ -342,12 +347,17 @@ ssize_t sock_dtls_send(sock_dtls_t *sock, sock_dtls_session_t *remote,
             DEBUG("sock_dtls: error initiating handshake\n");
             return -ENOMEM;
         }
-        else if (res > 0) {
+        else if (res > 0 && timeout != 0) {
             /* handshake initiated, wait until connected or timed out */
 
             msg_t msg;
             do {
-                res = xtimer_msg_receive_timeout(&msg, 3 * DTLS_HANDSHAKE_TIMEOUT);
+                uint32_t start = xtimer_now_usec();
+                res = xtimer_msg_receive_timeout(&msg, timeout);
+
+                if (timeout != SOCK_NO_TIMEOUT) {
+                    timeout = _update_timeout(start, timeout);
+                }
```

Comment:
```suggestion
                }
                if (timeout == 0) {
                    return -ETIMEOUT;
                }
```

Reply:
Refactored the existing loop condition a bit. Is it easier to understand this way?

```C
bool is_timed_out = false;
do {
    uint32_t start = xtimer_now_usec();
    res = xtimer_msg_receive_timeout(&msg, timeout);

    if (timeout != SOCK_NO_TIMEOUT) {
        timeout = _update_timeout(start, timeout);
        is_timed_out = (res < 0) || (timeout == 0);
    }
}
while (!is_timed_out && (msg.type != DTLS_EVENT_CONNECTED));
if (is_timed_out &&  (msg.type != DTLS_EVENT_CONNECTED)) {
    return -ETIMEDOUT;
}
```

Reply:
What happened to `return -EHOSTUNREACH` now?

Reply:
(but yes, in general I think it is more readable :-))

Reply:
Looking back, `EHOSTUNREACH` might be the wrong status code as it is for routes (no routes to host) but we don't know that here. We only know that `xtimer_msg_receive_timeout` has timed out.

Reply:
Mhhh ok.